lib/builderator/config/attributes.rb

```diff
@@ -0,0 +1,272 @@
+require 'forwardable'
+require 'json'
+
+require_relative './rash'
+
+module Builderator
+  module Config
+    ##
+    # Shared Attribute Mixin
+    ##
+    class Attributes
+      ##
+      # DSL Definition
+      ##
+      class << self
+        def attribute(attribute_name, default = nil, **options)
+          ##
+          # Helpers for Array-type attributes
+          ##
+          if options[:type] == :list
+            default = Array
+
+            ## Add an appender DSL method
+            define_method(options[:singular]) do |*args|
+              append_if_valid(attribute_name, args.flatten, default, options)
+            end if options.include?(:singular)
+          end
+
+          ## Getter/Setter
+          define_method(attribute_name) do |*arg|
+            arg.flatten!
+            arg = arg.first unless options[:type] == :list
+
+            set_or_return(attribute_name, arg, default, options)
+          end
+
+          ## Setter
+          define_method("#{attribute_name}=") do |arg|
+            set_if_valid(attribute_name, arg, options)
+          end
+        end
+
+        ## Add a method the DSL
+        def define(method_name, &block)
+          define_method(method_name, &block)
+        end
+
+        ##
+        # A Namespace is a singleton sub-node of the attribute-set
+        #
+        # e.g. `namespace :chef ...` maps to `attributes[:chef]` and adds a
+        # method `chef(&block)` to the DSL which is used as follows:
+        #
+        # ```
+        # chef do
+        #   run_list 'foo', 'bar'
+        #   ...
+        # end
+        # ```
+        #
+        # Multiple calls to the DSL method will are safe and will
+        # update the same sub-node.
+        ##
+        def namespace(namespace_name, &definition)
+          namespace_class = Namespace.create(namespace_name, &definition)
+
+          define_method(namespace_name) do |&block|
+            namespace_class.new(
+              @attributes[namespace_name],
+              :name => namespace_name, &block).compile
+          end
+        end
+
+        ##
+        # A Collection is a named-set of items in a sub-node of the attribute-set.
+        #
+        # Like Namespaces, Collections map to a top-level key, but they also have
+        # multiple second-order keys:
+        #
+        # e.g. `collection :vagrant ...` adds a DSL method
+        # `vagrant(name = :default, &block)` which maps to
+        # `attributes[:vagrant][<name>]`
+        #
+        # Multiple entities can be added to the collection by calling the DSL method
+        # with unique `name` arguments. Multiple calls to the DSL method with the
+        # same name argument will update the existing entity in place
```

Comment:
Can we add some tests around this behavior? I feels like the kind of thing future me could easily break while refactoring.